### PR TITLE
update table model

### PR DESCRIPTION
**Pull request by giovp**
as discussed in #158 

- [x] introduce mandatory `region` `region_key` and `instance_key`
- [x] rewrite concatenate following implementation
- [x] support string index io for shapes
- [x] change type of radius to `Union[float, ArrayLike]`

changes
- [x] the empty table case is not supported anymore, we can revert to this behaviour later but for now don't see priority

things leaving out of this PR as discussed in #158 
-  validation of type of categorical in for instances and instances in labels (should be done in spatialdata)
reason:
- not high priority and postponing to later work to consolidate spatialdata validation (e.g. we need a way to consolidate relationships between elements).

### CODE REVIEW

**Comment by giovp** [comment]
actually please don't merge as I forgot to add support for non-numeric indexes in geopandas io

File: spatialdata/_core/_spatialdata_ops.py
Comment (LucaMarconato) [inline spatialdata/_core/_spatialdata_ops.py:422]: I would keep this. Use case: we have a xenium and a visium dataset and we want to merge them keeping only the table of one of the two. To do so we can call `concatenate(..., omit_table=True)` and then do `sdata_concatenated.table = xenium.table`.

File: spatialdata/_core/_spatialdata_ops.py
Comment (LucaMarconato) [inline spatialdata/_core/_spatialdata_ops.py:329]: If all the `region_key` across the tables are identical, (`len(region_keys` == 1) I would use this value. So I would define `region_key` as `region_key: Optional[str] = None`.

File: spatialdata/_core/_spatialdata_ops.py
Comment (LucaMarconato) [inline spatialdata/_core/_spatialdata_ops.py:330]: Same here for what just said for `region_key`.

File: spatialdata/_core/models.py
Comment (LucaMarconato) [inline spatialdata/_core/models.py:682]: Isaac suggested to use `warn` instead of `logger.warning` across the repo: https://github.com/scverse/spatialdata/pull/163

File: spatialdata/_core/models.py
Comment (LucaMarconato) [inline spatialdata/_core/models.py:697]: `tqdm` has been removed from the dependencies of the package. But here I would remove the function `_sparse_matrix_from_assignment()`. It's a legacy from when we where representing points with `AnnData`.

File: spatialdata/_core/_spatialdata_ops.py
Comment (giovp) [reply spatialdata/_core/_spatialdata_ops.py:422]: what's the difference between manually adding the visium image and visium shapes to the xenium spatialdata?

Comment (LucaMarconato) [reply spatialdata/_core/_spatialdata_ops.py:422]: A big one: if you add an image to the xenium data, the new layer is saved to disk. On the contrary, `concatenate()` creates a new in-memory `sdata` object that is not backed (but the elements are backed from their respective `sdata` object)

Comment (giovp) [reply spatialdata/_core/_spatialdata_ops.py:422]: ok what about then simply passing one of the two `spatialdata` without the table? to me it just complicates the signature while the behavior is already impliclty implemetned

**Comment by giovp** [comment]
@LucaMarconato once this is merged, can #159 be closed?

File: spatialdata/_core/_spatialdata_ops.py
Comment (giovp) [reply spatialdata/_core/_spatialdata_ops.py:422]: > A big one: if you add an image to the xenium data, the new layer is saved to disk. On the contrary, concatenate() creates a new in-memory sdata object that is not backed (but the elements are backed from their respective sdata object)

this hidden behavior makes me also think we might want to be explicit about in memory/backed in concatenate

File: spatialdata/_io/write.py
Comment (LucaMarconato) [inline spatialdata/_io/write.py:272]: oh nice!

**Comment by LucaMarconato** [comment]
I'll have a quick double check of the tests that I added in https://github.com/scverse/spatialdata/pull/159, but now mostly is covered here. Btw I reviewed this PR, if you it's finished for me you can merge.

**Comment by giovp** [comment]
> I'll have a quick double check of the tests that I added in #159, but now mostly is covered here. Btw I reviewed this PR, if you it's finished for me you can merge.

I think it was mostly work on concatenate which has been re-added here.

File: spatialdata/_core/_spatialdata_ops.py
Comment (ivirshup) [inline spatialdata/_core/_spatialdata_ops.py:330]: Maybe something more like this?

```suggestion
    from copy import copy  # Should probably go up at the top

    tables_l = []
    for table_region_key, table_instance_key, table in zip(region_keys, instance_keys, tables):
        rename_dict = {}
        if table_region_key != region_key:
            rename_dict[table_region_key] = region_key
        if table_instance_key != instance_key:
            rename_dict[table_instance_key] = instance_key
        if len(rename_dict) > 0:
            table = copy(table)  # Shallow copy
            table.obs = table.obs.rename(columns=rename_dict, copy=False)
        tables_l.append(table)

    merged_table = ad.concat(tables_l, **kwargs)
```

* I found the current version took a little effort to figure out. If you want to keep it, it could use a comment explaining that it's making the column names consistent
* This should be more performant, as it doesn't require the subsetting, and the shallow copy should be close to free

The result is slightly different. The order of the observations in tables is preserved, and the result couldn't include the original region or instance keys. If you wanted the possibility of preserving all the original columns in an `outer` join, you could add copies of the instance and region key columns instead of renaming them.

File: spatialdata/_core/_spatialdata_ops.py
Comment (ivirshup) [inline spatialdata/_core/_spatialdata_ops.py:375]: Could the Exeptions bellow this be `KeyError`s or `ValueError`s?

Also these checks are quite cheap, can they happen before the potentially expensive table concatenation?

File: spatialdata/_core/_spatialdata_ops.py
Comment (ivirshup) [inline spatialdata/_core/_spatialdata_ops.py:369]: Can we remove the special casing for the "one table passed" case?

It's weird to me that it:

* A reference to the object that was passed in. So any modification of the result would modify the original, breaking the semantics of concatenation.
* The "renaming" properties of `region_key` and `instance_key` arguments are not respected.

Tests for this case would be great.

**Review by kevinyamauchi — APPROVED** [review]
LGTM! Thanks!

File: spatialdata/_core/_spatialdata_ops.py
Comment (ivirshup) [reply spatialdata/_core/_spatialdata_ops.py:422]: I would remove this. I don't think the feature being described fits the definition of concatenation. Maybe this is more of a `"merge"`?

Also, is super easy for a user to do themselves in the meantime.

<details>
<summary> example </summary>

```python
from functools import reduce
from operator import or_

sdata_w_table = ...
other_sdatas: list[SpatialData]

sdatas = [sdata_w_table, *other_sdatas]

new_sdata = SpatialData(
    table=sdata_w_table.table,
    images=reduce(or_, (sdata.images for sdata in sdatas)),
    labels=reduce(or_, (sdata.labels for sdata in sdatas)),
    shapes=reduce(or_, (sdata.shapes for sdata in sdatas)),
    points=reduce(or_, (sdata.points for sdata in sdatas)),
)
```

</details>

File: spatialdata/_core/models.py
Comment (ivirshup) [inline spatialdata/_core/models.py:644]: Shouldn't the type of the argument class already be included from the type signature?

```suggestion
```

File: spatialdata/_core/models.py
Comment (ivirshup) [inline spatialdata/_core/models.py:677]: Is this necessary?

If mypy really can't figure out the previous line, could you change it to `region_: list = ...` instead?

```suggestion
```

File: tests/_core/test_spatialdata_operations.py
Comment (ivirshup) [inline tests/_core/test_spatialdata_operations.py:146]: Why was this removed?

File: tests/_core/test_spatialdata_operations.py
Comment (ivirshup) [inline tests/_core/test_spatialdata_operations.py:136]: Why is there an assertion in here?

The assertion is never executed, so I'm not seeing the point in including it.

**Review by ivirshup — COMMENTED** [review]
Mostly good, a few suggestions and minor things.

File: spatialdata/_core/_spatialdata_ops.py
Comment (LucaMarconato) [reply spatialdata/_core/_spatialdata_ops.py:369]: You are right

File: tests/_core/test_spatialdata_operations.py
Comment (LucaMarconato) [reply tests/_core/test_spatialdata_operations.py:136]: my mistake

**Comment by LucaMarconato** [comment]
Answering to the comment re `omit_table`, https://github.com/scverse/spatialdata/pull/164#discussion_r1124456600. Maybe a `merge` would be better here. Not sure. I would then remove it for now and then we see how to provide this functionality. For the moment in the Xenium + Visium notebook we can either use the `reduce()` approach (or equivalent) or never merge the two `sdata` objects (we can do most/all the operations keeping them separate actually).

**Comment by ivirshup** [comment]
@giovp, actually one other (maybe major) point.

What happens if (for any of the tables) instance key is `None`?

**Comment by LucaMarconato** [comment]
> What happens if (for any of the tables) instance key is None?
I see that this is not possible for a table created with the parser (see here https://github.com/scverse/spatialdata/blob/80dc1024d397a06120f9bd5b27237b9772a85997/spatialdata/_core/models.py#LL685-L686C65), but a table with `instance_key = None` would pass validation. I would update the validation to catch this case and raise an exception.

File: spatialdata/_core/_spatialdata_ops.py
Comment (giovp) [reply spatialdata/_core/_spatialdata_ops.py:366]: you mean this case?
```
if len(sdatas) == 1:
        return sdatas[0]
```

File: spatialdata/_core/models.py
Comment (giovp) [reply spatialdata/_core/models.py:644]: ok, think it'd be nice to also implement a docstring processor likw we have in squidpy

File: tests/_core/test_spatialdata_operations.py
Comment (giovp) [reply tests/_core/test_spatialdata_operations.py:146]: redundant with the new one

**Comment by giovp** [comment]
thanks all for feedback! I accepted all suggestions and made couple of modifications
- explicitly check for regions in uns and obs, e.g. (spotted one bug in filter coordinate system @LucaMarconato but fixed it)

```python
set(adata.uns["spatialdata_attrs"]["regions"] ) == set(adata.obs.region_key.unqiue())
# True
```

- explicitly check for all spatialdata_attrs to be valid (e.g. value of `instance_key` cannot be None).
- accepted code suggestion by @ivirshup but added additional check on unique region names that suggested version missed, e.g.:
```python
sdata0 = SpatialData(labels={"sample1":..}, table}
sdata1 = SpatialData(labels={"sample1":..}, table}
concatenate(sdata0, sdata1)
# now throws error, as there is no way to understand to which obs refers which region.
```
the latter is something that ideally we will change, but it'd require keeping track of region names and rename them in the resulting spatialdata object. If tests pass I'll go on and merge

**Pull request merged**
